connectors/sources/box.py

```diff
@@ -0,0 +1,445 @@
+        self.is_enterprise = configuration["is_enterprise"]
+        self._http_session = aiohttp.ClientSession(
+            base_url="https://api.box.com", raise_for_status=True
+        )
+        self.token = AccessToken(
+            configuration=configuration, http_session=self._http_session
+        )
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    async def _put_to_sleep(self, exception):
+        retry_after = int(exception.headers.get("retry-after", 5))
+        self._logger.debug(
+            f"Connector will attempt to retry after {retry_after} seconds."
+        )
+        await self._sleeps.sleep(retry_after)
+        raise Exception("Rate limit exceeded.")
+
+    async def _handle_client_errors(self, exception):
+        match exception.status:
+            case 401:
+                await self.token._set_access_token()
+                raise
+            case 429:
+                await self._put_to_sleep(exception=exception)
+            case 404:
+                raise NotFound(f"Resource Not Found. Error: {exception}")
+            case _:
+                raise
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+        skipped_exceptions=NotFound,
+    )
+    async def get(self, url, headers, params=None):
+        try:
+            access_token = await self.token.get()
+            headers.update({"Authorization": f"Bearer {access_token}"})
+            return await self._http_session.get(url=url, headers=headers, params=params)
+        except ClientResponseError as exception:
+            await self._handle_client_errors(exception=exception)
+        except Exception:
+            raise
+
+    async def paginated_call(self, url, params, headers):
+        try:
+            offset = 0
+            while True:
+                params.update({"offset": offset, "limit": FETCH_LIMIT})
+                response = await self.get(url=url, headers=headers, params=params)
+                json_response = await response.json()
+                total_count = json_response.get("total_count")
+                for doc in json_response.get("entries"):
+                    yield doc
+                if offset >= total_count:
+                    break
+                offset += FETCH_LIMIT
+        except Exception:
+            raise
+
+    async def ping(self):
+        await self.get(url=ENDPOINTS["PING"], headers={})
+
+    async def close(self):
+        self._sleeps.cancel()
+        await self._http_session.close()
+
+
+class BoxDataSource(BaseDataSource):
+    name = "Box"
+    service_type = "box"
+
+    def __init__(self, configuration):
+        super().__init__(configuration=configuration)
+        self.configuration = configuration
+        self.tasks = 0
+        self.is_enterprise = configuration["is_enterprise"]
+        self.queue = MemQueue(maxmemsize=QUEUE_MEM_SIZE, refresh_timeout=120)
```

Comment:
What is the need to have a queue?

Reply:
We are fetching files in a recursive manner in the `_fetch_files_folder`. So, without fetchers, it's fetching files in a synchronized manner only. But with the fetchers, when the `_fetch_files_folder` function is called, we put it into the fetchers, and the fetchers add that document to the queue, and parallelly consumer function will yield the data from the queue.